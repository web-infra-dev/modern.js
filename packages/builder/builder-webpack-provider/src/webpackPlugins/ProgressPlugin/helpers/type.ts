import type { ForegroundColor as Color } from '@modern-js/utils/compiled/chalk';

export type Props = {
  total: number;
  current: number;
  color: typeof Color;
  bgColor: typeof Color;
  char: string;
  width: number;
  buildIcon: string;
  finishIcon: string;
  finishInfo: string;
  message: string;
  done: boolean;
  messageWidth: number;
  spaceWidth: number;
  messageColor: typeof Color;
  id: string;
  maxIdLen: number;
};

export type BusOption = {
  state: Props[];
};
