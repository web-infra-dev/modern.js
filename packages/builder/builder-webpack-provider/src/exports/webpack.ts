// eslint-disable-next-line @typescript-eslint/no-restricted-imports
import webpack from 'webpack';

export default webpack;
