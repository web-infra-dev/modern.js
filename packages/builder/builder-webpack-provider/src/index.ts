export { builderWebpackProvider } from './provider';
export type { BuilderWebpackProvider } from './provider';
export { webpackBuild } from './core/build';
export type { WebpackBuildError } from './core/build';
export { createDefaultConfig } from './config/defaults';
export type {
  BuilderPlugin,
  BuilderPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  OutputConfig,
  SourceConfig,
  BuilderConfig,
  SecurityConfig,
  NormalizedConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Hook Callback Types
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSLoaderOptions,
  CSSExtractOptions,
  LessLoaderOptions,
  SassLoaderOptions,
  HTMLPluginOptions,
  StyleLoaderOptions,
  AutoprefixerOptions,
  TerserPluginOptions,
  PostCSSLoaderOptions,
} from './types';
