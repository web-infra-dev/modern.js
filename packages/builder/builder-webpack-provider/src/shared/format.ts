import type { webpack } from '../types';
import chalk from '@modern-js/utils/chalk';

export async function formatWebpackStats(
  stats: webpack.Stats | webpack.MultiStats,
  showWarnings = true,
) {
  const statsData = stats.toJson({
    preset: 'errors-warnings',
  });

  const { formatWebpackMessages } = await import('@modern-js/utils');
  const { errors, warnings } = formatWebpackMessages(statsData);

  if (errors.length) {
    const title = chalk.red.bold(`Compile Error: \n`);
    return {
      message: `${title}${errors.join('\n\n')}\n`,
      level: 'error',
    };
  }

  // always show warnings in tty mode
  if (warnings.length && (showWarnings || process.stdout.isTTY)) {
    const title = chalk.yellow.bold(`Compile Warning: \n`);
    return {
      message: `${title}${`${warnings.join('\n\n')}\n`}`,
      level: 'warning',
    };
  }

  return {};
}
