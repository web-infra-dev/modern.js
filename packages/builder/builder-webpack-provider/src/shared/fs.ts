import { join } from 'path';
import {
  CSS_DIST_DIR,
  FONT_DIST_DIR,
  HTML_DIST_DIR,
  IMAGE_DIST_DIR,
  JS_DIST_DIR,
  MEDIA_DIST_DIR,
  ROOT_DIST_DIR,
  SVG_DIST_DIR,
} from '@modern-js/builder-shared';
import type { BuilderConfig, DistPathConfig, FilenameConfig } from '../types';

export const getCompiledPath = (packageName: string) =>
  join(__dirname, '../../compiled', packageName);

export const getDistPath = (
  config: BuilderConfig,
  type: keyof DistPathConfig,
): string => {
  const { distPath = {} } = config.output || {};

  switch (type) {
    case 'js':
      return distPath.js ?? JS_DIST_DIR;
    case 'css':
      return distPath.css ?? CSS_DIST_DIR;
    case 'svg':
      return distPath.svg ?? SVG_DIST_DIR;
    case 'font':
      return distPath.font ?? FONT_DIST_DIR;
    case 'html':
      return distPath.html ?? HTML_DIST_DIR;
    case 'media':
      return distPath.media ?? MEDIA_DIST_DIR;
    case 'root':
      return distPath.root ?? ROOT_DIST_DIR;
    case 'image':
      return distPath.image ?? IMAGE_DIST_DIR;
    default:
      throw new Error(`unknown key ${type} in "output.distPath"`);
  }
};

export const getFilename = (
  config: BuilderConfig,
  type: keyof FilenameConfig,
  isProd: boolean,
) => {
  const { filename = {} } = config.output || {};
  const useHash = isProd && !config.output?.disableFilenameHash;
  const hash = useHash ? '.[contenthash:8]' : '';

  switch (type) {
    case 'js':
      return filename.js ?? `[name]${hash}.js`;
    case 'css':
      return filename.css ?? `[name]${hash}.css`;
    case 'svg':
      return filename.svg ?? `[name]${hash}.svg`;
    case 'font':
      return filename.font ?? `[name]${hash}[ext]`;
    case 'image':
      return filename.image ?? `[name]${hash}[ext]`;
    case 'media':
      return filename.media ?? `[name]${hash}[ext]`;
    default:
      throw new Error(`unknown key ${type} in "output.filename"`);
  }
};
