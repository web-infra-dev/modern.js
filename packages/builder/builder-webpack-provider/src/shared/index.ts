export * from './fs';
export * from './utils';
export * from './format';
