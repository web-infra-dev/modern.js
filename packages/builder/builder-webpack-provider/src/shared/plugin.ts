import { BuilderPlugin } from '../types';
import { awaitableGetter, Plugins } from '@modern-js/builder-shared';

export const applyMinimalPlugins = (plugins: Plugins) =>
  awaitableGetter<BuilderPlugin>([
    import('../plugins/basic').then(m => m.PluginBasic()),
    plugins.entry(),
    import('../plugins/cache').then(m => m.PluginCache()),
    plugins.target(),
    import('../plugins/output').then(m => m.PluginOutput()),
    plugins.devtool(),
    import('../plugins/resolve').then(m => m.PluginResolve()),
  ]);

export const applyBasicPlugins = (plugins: Plugins) =>
  awaitableGetter<BuilderPlugin>([
    ...applyMinimalPlugins(plugins).promises,
    import('../plugins/copy').then(m => m.PluginCopy()),
    import('../plugins/html').then(m => m.PluginHtml()),
    import('../plugins/image').then(m => m.PluginImage()),
    import('../plugins/define').then(m => m.PluginDefine()),
    import('../plugins/tsLoader').then(m => m.PluginTsLoader()),
    import('../plugins/babel').then(m => m.PluginBabel()),
    import('../plugins/css').then(m => m.PluginCss()),
    import('../plugins/sass').then(m => m.PluginSass()),
    import('../plugins/less').then(m => m.PluginLess()),
    import('../plugins/react').then(m => m.PluginReact()),
  ]);

export const applyDefaultPlugins = (plugins: Plugins) =>
  awaitableGetter<BuilderPlugin>([
    ...applyMinimalPlugins(plugins).promises,
    plugins.fileSize(),
    plugins.cleanOutput(),
    import('../plugins/hmr').then(m => m.PluginHMR()),
    import('../plugins/svg').then(m => m.PluginSvg()),
    import('../plugins/pug').then(m => m.PluginPug()),
    import('../plugins/copy').then(m => m.PluginCopy()),
    import('../plugins/font').then(m => m.PluginFont()),
    import('../plugins/html').then(m => m.PluginHtml()),
    import('../plugins/image').then(m => m.PluginImage()),
    import('../plugins/media').then(m => m.PluginMedia()),
    import('../plugins/moment').then(m => m.PluginMoment()),
    import('../plugins/define').then(m => m.PluginDefine()),
    import('../plugins/progress').then(m => m.PluginProgress()),
    import('../plugins/minimize').then(m => m.PluginMinimize()),
    import('../plugins/manifest').then(m => m.PluginManifest()),
    import('../plugins/moduleScopes').then(m => m.PluginModuleScopes()),
    import('../plugins/tsLoader').then(m => m.PluginTsLoader()),
    import('../plugins/babel').then(m => m.PluginBabel()),
    import('../plugins/tsChecker').then(m => m.PluginTsChecker()),
    import('../plugins/css').then(m => m.PluginCss()),
    import('../plugins/sass').then(m => m.PluginSass()),
    import('../plugins/less').then(m => m.PluginLess()),
    import('../plugins/rem').then(m => m.PluginRem()),
    import('../plugins/react').then(m => m.PluginReact()),
    import('../plugins/bundleAnalyzer').then(m => m.PluginBundleAnalyzer()),
    import('../plugins/toml').then(m => m.PluginToml()),
    import('../plugins/yaml').then(m => m.PluginYaml()),
    import('../plugins/splitChunks').then(m => m.PluginSplitChunks()),
    import('../plugins/inspector').then(m => m.PluginInspector()),
    import('../plugins/sri').then(m => m.PluginSRI()),
    plugins.startUrl(),
    import('../plugins/inlineChunk').then(m => m.PluginInlineChunk()),
    import('../plugins/assetsRetry').then(m => m.PluginAssetsRetry()),
    import('../plugins/externals').then(m => m.PluginExternals()),
    import('../plugins/performance').then(m => m.PluginPerformance()),
    import('../plugins/lazyCompilation').then(m => m.PluginLazyCompilation()),
    import('../plugins/fallback').then(m => m.PluginFallback()), // fallback should be the last plugin
  ]);
