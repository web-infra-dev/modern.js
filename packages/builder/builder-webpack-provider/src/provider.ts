import {
  createPublicContext,
  type BuilderProvider,
} from '@modern-js/builder-shared';
import { createContext } from './core/createContext';
import { applyDefaultPlugins } from './shared/plugin';
import { BuilderConfig } from './types';
import { initConfigs } from './core/initConfigs';

export function builderWebpackProvider({
  builderConfig,
}: {
  builderConfig: BuilderConfig;
}): BuilderProvider {
  return async ({ pluginStore, builderOptions }) => {
    const context = await createContext(builderOptions, builderConfig);

    return {
      bundler: 'webpack',

      publicContext: createPublicContext(context),

      async applyDefaultPlugins() {
        pluginStore.addPlugins(await applyDefaultPlugins());
      },

      async createCompiler({ watch } = {}) {
        const { createCompiler } = await import('./core/createCompiler');
        const { webpackConfigs } = await initConfigs({
          context,
          pluginStore,
          builderOptions,
        });
        return createCompiler({ watch, context, webpackConfigs });
      },

      async startDevServer(options) {
        const { startDevServer } = await import('./core/startDevServer');
        return startDevServer(
          { context, pluginStore, builderOptions },
          options,
        );
      },

      async build(options) {
        const { build: buildImpl, webpackBuild } = await import('./core/build');
        return buildImpl(
          { context, pluginStore, builderOptions },
          options,
          webpackBuild,
        );
      },

      async inspectConfig(inspectOptions) {
        const { inspectConfig } = await import('./core/inspectConfig');
        return await inspectConfig({
          context,
          pluginStore,
          builderOptions,
          inspectOptions,
        });
      },
    };
  };
}
