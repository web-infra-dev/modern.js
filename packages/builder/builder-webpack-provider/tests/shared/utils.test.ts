import webpack from 'webpack';
import { expect, describe, it } from 'vitest';
import { stringifyConfig } from '../../src/shared';
import type { BuilderConfig, WebpackConfig } from '../../src/types';

describe('stringifyConfig', () => {
  it('should stringify webpack config correctly', async () => {
    const { DefinePlugin } = webpack;
    const webpackConfig: WebpackConfig = {
      mode: 'development',
      plugins: [new DefinePlugin({ foo: 'bar' })],
    };

    expect(await stringifyConfig(webpackConfig)).toMatchSnapshot();
  });

  it('should stringify builder config correctly', async () => {
    const builderConfig: BuilderConfig = {
      tools: {
        webpackChain(chain) {
          chain.devtool('eval');
        },
      },
    };

    expect(await stringifyConfig(builderConfig)).toMatchSnapshot();
  });
});
