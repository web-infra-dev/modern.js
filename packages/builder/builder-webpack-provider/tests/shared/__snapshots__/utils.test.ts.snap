// Vitest Snapshot v1

exports[`stringifyConfig > should stringify builder config correctly 1`] = `
"{
  tools: {
    webpackChain(chain) {
      chain.devtool(\\"eval\\");
    }
  }
}"
`;

exports[`stringifyConfig > should stringify webpack config correctly 1`] = `
"{
  mode: 'development',
  plugins: [
    {
      definitions: {
        foo: 'bar'
      }
    }
  ]
}"
`;

exports[`stringifyConfig > should stringify webpack config with verbose option correctly 1`] = `
"{
  mode: 'development',
  plugins: [
    {
      definitions: {
        foo: 'bar',
        baz() {
          const a = 1;
          return a;
        }
      }
    }
  ]
}"
`;
